### ISSUE TITLE
Module compiled with Swift 4.0 cannot be imported in Swift 4.0.2: 

### ISSUE THREAD

**Issue by sLm-s** [layoutBox/FlexLayout#34]
Hi @lucdion. Gives an error in Xcode 9.1

**Comment by sLm-s** [comment]
And there is one more question. How to create such a layout, as in the picture? There are two texts that are always located at the edges opposite each other and between them the line.

![simulator screen shot - iphone 8 - 2017-11-03 at 12 49 13](https://user-images.githubusercontent.com/20534572/32362946-f624c5f2-c096-11e7-8e19-5ac33b16d137.png)

**Comment by sLm-s** [comment]
@lucdion I need your help. Error -- Module compiled with Swift 4.0 cannot be imported in Swift 4.0.2 
in xCode 9.1

**Comment by lucdion** [comment]
@sLm-s if you use carthage, delete your `Carthage` directory and reintall carthage modules. If you use pod, be sure to clean up everything before rebuilding (even sometimes the derived data directory).

**Comment by lucdion** [comment]
For your problem above, sorry I didn't have the time yet to test it using flexbox. 
Using PinLayout it would be:
```
redText.pin.left().vCenter().fitSize()
purpleText.pin.right().vCenter().fitSize()
lineView.pin.after(of: redText).before(of: purpleText).vCenter().marginHorizontal(10)
```
I'll test it using FlexLayout later today. Thanks

**Comment by sLm-s** [comment]
Thank you so much. I'll wait for the FlexLayut example. I do not use PinLayout.

**Comment by lucdion** [comment]
@sLm-s, here is the solution to your layout:

```
           flex.addItem().direction(.row).padding(2, 6, 2, 6).backgroundColor(.cyan).define({ (flex) in
                flex.addItem(redLabel)
                flex.addItem(lineSeparator).height(3).grow(1).marginHorizontal(6).alignSelf(.center)
                flex.addItem(purpleLabel)
            })
``` 

![screen shot 2017-11-06 at 10 34 25 am](https://user-images.githubusercontent.com/14981341/32448897-272a68fc-c2de-11e7-8a2f-a2a5da510f65.png)

**Comment by sLm-s** [comment]
@lucdion Is it possible to change the width of an object, for example, if it was initially set to zero but after calculations 100. Excuse me for my obsession.

**Comment by lucdion** [comment]
Yes, it should work if you do `myView.flex.width(100)` and then relayout the flex container.

**Comment by sLm-s** [comment]
override init(frame: CGRect) {
        super.init(frame: frame)
        
        contentView.flex.addItem().direction(.row).paddingVertical(1).define({ (flex) in
            flex.addItem(imageStar1).width(widthStar).marginLeft(1)
            flex.addItem(imageStar2).width(widthStar).marginLeft(1)
            flex.addItem(imageStar3).width(widthStar).marginLeft(1)
            flex.addItem(imageStar4).width(widthStar).marginLeft(1)
            flex.addItem(imageStar5).width(widthStar).marginLeft(1)
                 flex.addItem(fonLineView).height(heightLine).grow(1).marginHorizontal(8).alignSelf(.center).define({ (flex) in
                flex.addItem(lineView).grow(1)
            })
            flex.addItem(countMark).width(25)
        })


    }


But then I want to install lineView.width = 100
lineView.flex.width(100)  this does not work

lineView.frame.size.width = 100 it works

**Comment by lucdion** [comment]
Setting directly `lineView.frame.size.width` will work until the flex container will be relayouted by `contentView.flex.layout()`. After `lineView.flex.width(100)` (you may have to call `lineView.flex.markAsDirty()` too, I'm not sure) you must relayout the flex container using `contentView.flex.layout()`.

**Comment by sLm-s** [comment]
@lucdion  Thank you so much. It works.
lineView.flex.width(width)
contentView.flex.layout()

**Comment by sLm-s** [comment]
There is one more problem. FlexLayout incorrectly determines the size of the button.
And if there is a lot of text in the button, it does not change in size.

![simulator screen shot - iphone 8 - 2017-11-07 at 10 09 19](https://user-images.githubusercontent.com/20534572/32476877-82c6cbd0-c3a4-11e7-8a1c-2d0329f6cf94.png)
![simulator screen shot - iphone 8 - 2017-11-07 at 10 09 33](https://user-images.githubusercontent.com/20534572/32476878-82eb4d34-c3a4-11e7-8850-eda19057396c.png)

**Comment by sLm-s** [comment]
I cannot fix the error

Module compiled with Swift 4.0 cannot be imported in Swift 4.0.2: /Users/slm/Documents/Project/Flip/Carthage/Build/iOS/FlexLayout.framework/Modules/FlexLayout.swiftmodule/x86_64.swiftmodule

**Comment by lucdion** [comment]
@sLm-s I have fixed the carthage compilation, sorry for that.

**Comment by lucdion** [comment]
@sLm-s for your button issue. It looks like the UIButton sizeThatFits always returns an height matching one text line. If you call this line in `layoutSubviews`
```
let size = button.sizeThatFits(CGSize(width: frame.width, height: .greatestFiniteMagnitude))
```
The size.height is always the same value, even if the text is longer.

It looks like if you want to have a button with a multiline label, you'll have to subclass UIButton and implement the sizeThatFit method.

**Comment by sLm-s** [comment]
@lucdion Thank you for your help and your work.

**Comment by sLm-s** [comment]
Hi @lucdion. Whether it is possible to make the fixed size of the container, and the FlexLayout calculated the size of all objects this container ?

**Comment by sLm-s** [comment]
Other question: whether it is possible to make the container transparent and in it objects not transparent. Or create transparent background for a container

**Comment by lucdion** [comment]
The container can be transparent by setting its background to UIColor.clear.
I'm not sure to understand your question related to "fixed size of the container", can you rephrase it?

**Comment by sLm-s** [comment]
Here is an approximate problem.

![simulator screen shot - iphone 8 - 2017-11-10 at 09 29 52](https://user-images.githubusercontent.com/20534572/32642277-51228e92-c5fd-11e7-8b13-deca2d676774.png)
![simulator screen shot - iphone 8 - 2017-11-10 at 09 39 51](https://user-images.githubusercontent.com/20534572/32642278-51974354-c5fd-11e7-9704-0db239c36f9f.png)

**Comment by sLm-s** [comment]
Hi @lucdion . Is it possible to determine the coordinates of the object inside the container FlexLayout ?

**Comment by lucdion** [comment]
If you keep a reference on your object (view), you can always retrieve its `frame` property after the container has been layouted (ex: `myView.frame.origin.x`, `myView.frame.origin.y`)

**Comment by lucdion** [comment]
For tips for UICollectionView, you can check how this has been achieved in the UICollectionView sample.

**Comment by sLm-s** [comment]
How to be if you need a specific size and not objects inside the cell. For example, the height is 320 and the height of all objects is 280 then I need these objects to stretch to the entire height of the cell 320 more precisely those that can for example UILabel

**Comment by sLm-s** [comment]
Created shoot there are several kinds of collections placed in it. If the turning screen fires the Villtransition method. But the problem is when you return to the previous screen the layout changes its dimensions in your eyes and it doesn't look very nice.

override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)

        coordinator.animate(alongsideTransition: { (context) in
            self.layout()
        }) { (completed) in
//            self.layout()

        }
        

        
    }

**Comment by lucdion** [comment]
@sLm-s to stretch items you must specify the [grow value](https://github.com/layoutBox/FlexLayout#grow) for at least one of the container's children. For example, if you specify a grow value for all container's children, the available space will be equally distributed. By default the grow value is 0, which don't expand.

Sometimes you also need to specify the `shrink` value if the container is smaller than its children size.

**Comment by sLm-s** [comment]
@lucdion Unfortunately this does not help. Here are the screenshots.
I open the application the first screen appears, I rotate the in landscape. And I turn to a new screen, then I turn over to the portrait view and go back. After this, the calculation of the layout is not correct. But the second passes and the layout takes a normal form

![simulator screen shot - ipad pro 9 7-inch - 2017-11-15 at 08 51 52](https://user-images.githubusercontent.com/20534572/32817309-229b2726-c9e7-11e7-9f80-dd745294d1e1.png)
![simulator screen shot - ipad pro 9 7-inch - 2017-11-15 at 08 54 41](https://user-images.githubusercontent.com/20534572/32817310-22bd3064-c9e7-11e7-91e2-bc2c82614b9e.png)
![simulator screen shot - ipad pro 9 7-inch - 2017-11-15 at 09 10 17](https://user-images.githubusercontent.com/20534572/32817311-22e46fe4-c9e7-11e7-9283-704413c349c0.png)

**Comment by sLm-s** [comment]
Your FlexLayout helps me very cool. One problem a lot of questions)). I'm not an experienced yet developer((.

**Comment by lucdion** [comment]
Hi @sLm-s, could you share  me your whole project, or at least extract those screens that have issues. I could look at it.

**Comment by sLm-s** [comment]
I do not even know how to show you the part of the project so you can understand it and. I filled in the code in which I create a layout with the help of UICroll. There are several types of Collections in it. When I click on a cell, I switch to the same controller but with a different layout. And when I start to rotate the screen the layout is well updated but if I go back to the previous controller then the layout is not updated normally. Sorry for my bad English

[ScrollCatalog.zip](https://github.com/layoutBox/FlexLayout/files/1474588/ScrollCatalog.zip)

**Comment by lucdion** [comment]
@sLm-s if you completely comment the code in `viewWillTransition` in ScrollCatalog, and simply depend on the `viewDidLayoutSubviews` method, does it work correctly?

Also to clean up a little bit your layouting code, you should probably move all your layout code from `viewDidLayoutSubviews` into your `layout` method. This way all layout related code would be at a single place.

**Comment by sLm-s** [comment]
In viewWillTransition uses only layout() the rest is not important

override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        super.viewWillTransition(to: size, with: coordinator)

        coordinator.animate(alongsideTransition: { [weak self] (context) in
            
            guard let strongSelf = self else { return }
            
            
            strongSelf.layout()
            
        }) { (completed) in

        }


    }

**Comment by sLm-s** [comment]
In viewDidLayoutSubviews looks like this. the rest are additional chips

override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()

        
        scrollView.frame = view.frame
        rootFlexContainer.frame.size.width = view.frame.width
        
        rootFlexContainer.flex.layout(mode: .adjustHeight)
        
        scrollView.contentSize = rootFlexContainer.frame.size
        
       
            }
        }
        
    }

**Comment by sLm-s** [comment]
In file parsScrollCatalog.swift created the whole layout

**Comment by sLm-s** [comment]
viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) 
Does not work for layout recalculation?

**Comment by sLm-s** [comment]
Hi @lucdion . Thank you for your help. I figured out why the layout was not working correctly.

**Comment by lucdion** [comment]
No problem, happy to try to help you!
